### DISCUSSION TITLE
Tft e3 v3.0 flipped letters

### DISCUSSION THREAD

**Discussion by Deadmanfpv** [teeminus/NoTouchScreenFirmware#57] [Q&A]
hello. I just updated my BTT tft35 e3 v3.0 to this firmware. Im using an e3 mini v2 mainboard.  It was successful however all of the letters are flipped upside down for some reason. I used the proper bin file and just copy/pasted the known working config for my setup. any answers as to how i can fix this?

**Comment by teeminus** [comment]
Can you please post a picture of your TFT?

**Reply by TeStevens315** [reply]
I have the same issue I just posted it in issues, not sure if that is the correct place for it or not.

Here is a picture of my screen.
![PXL_20211101_221552906 MP](https://user-images.githubusercontent.com/20211492/139758504-9234f4a3-133e-48c5-8678-55462fa5428a.jpg)

**Reply by teeminus** [reply]
Thats not an actual issue. Some printer manufacturers flip the screen to mount it 180° rotated (See https://github.com/teeminus/NoTouchScreenFirmware/issues/8). You can manuall recompile the firmware with the following line commented in: https://github.com/teeminus/NoTouchScreenFirmware/blob/63a382f294762900353005e41adb902b55045938/src/User/features.h#L11

**Reply by TeStevens315** [reply]
Thanks for the pointer, and the quick response. I am not much of a programmer and sometimes I need a little help to figure out what to modify.

So success in that my screen is now rotated but, now it is still mirrored.   I poked around in the code but I can't find an obvious way to mirror the text.   I think I might need another hint.

Thanks for your help.
Troy
![PXL_20211102_163158609 MP](https://user-images.githubusercontent.com/20211492/139909429-fb26c62b-0f92-4274-bda0-2143414878cd.jpg)

**Reply by teeminus** [reply]
I will have a look in the code to see how we can fix the issue.

**Reply by teeminus** [reply]
Checkout the latest v1.3.1 release and compile it manually with the following line commented in: https://github.com/teeminus/NoTouchScreenFirmware/blob/9dbcf07111100c61d75b52dedad14d8b33871b44/src/User/features.h#L14

Please note that you need to unset the `LCD_ROTATE_180` define before (if it is enabled in your checkout).

**Reply by TeStevens315** [reply]
Well, it flipped the screen to the correct orientation, however, now it doesn't show the interface past "ST7920Emulator ready"   

If I hold down the encoder button I can get past that but I just have a blank screen with "NoTouchFw v1.31" along the top.   

I reflashed 1.3 just to make sure I didn't mess up my config in Klipper and I got the interface back but with the flipped lettering.

For reference here is my config line for the display:
[display]
lcd_type: st7920
cs_pin: EXP1_4
sclk_pin: EXP1_5
sid_pin: EXP1_3
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2


I had also tried another config from another discussion that had different pin numbers than mine even though they had an skr2 w/ tft35 e3 display.   It didn't work either.   My above config works with the V1.3 build.

Thanks for your help
Troy

**Reply by teeminus** [reply]
If the klipper config has worked for v1.3 it should work for v1.3.1 as well. The difference between the two version is just the "improved"/extended mirroring functionality.

I just flashed my TFT24v1.1 with the vertical mirroring enabled and it worked as expected...

Have you tried the precompiled v1.3.1 version I uploaded with the release? Does this firmware works (besides beeing mirrored on your screen)?

When stuck at the "ST7920Emulator ready", is the header line below/around the "NoTouchFw v1.3.1" title drawn correctly (from left side of screen to the title and from the end of the title to the right side of the screen)?

If you hold down the encoder button for 5 seconds, have you tried restarting klipper? The long press of the button re-initializes the SPI-Interface of the screen, might be worth a try.

You can give my precompiled TFT24v1.1 firmware with enabled vertical mirroring a try if you like: [BIGTREE_TFT24_V1.1.26.x.zip](https://github.com/teeminus/NoTouchScreenFirmware/files/7470132/BIGTREE_TFT24_V1.1.26.x.zip)

**Reply by TeStevens315** [reply]
I got it working somewhat.   I played around with 1.3.1 and it turns out after resetting the SPI-interface, I have to mess with the encoder knob to get into the interface.    

It seems a little transient.    After resetting the SPI it needs a short delay.   Then I rotate the encoder knob and I get a different random letter or number or sometimes the ..  icon.  After I press the encoder knob I can reliably get to the main screen.

Anyways that works for me for now, I went from thinking I wouldn't be able to use my new screen for klipper to being able to use it.   This will be my first kipper build so I am not sure how much I will even use the screen that much, vs using klipper directly from my raspberry pi touch screen.

Thanks for all the work you have done getting this to work on our screens.

Troy

**Reply by teeminus** [reply]
Glad to hear that your TFT works now - kind of... 😅

By the way, have you tried the emulated_st7920 Klipper display driver?
